test/integration/endpointslice/endpointslice_test.go

```diff
@@ -0,0 +1,305 @@
+	go epController.Run(tCtx, 1)
+
+	ns := framework.CreateNamespaceOrDie(client, "test-endpoints-semantics", t)
+	defer framework.DeleteNamespaceOrDie(client, ns, t)
+
+	svc := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "test-123",
+			Labels: map[string]string{
+				"app": "test",
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Type:      corev1.ServiceTypeClusterIP,
+			ClusterIP: corev1.ClusterIPNone, // Headless service
+			Selector: map[string]string{
+				"app": "test",
+			},
+			Ports: []corev1.ServicePort{
+				{
+					Port:       80,
+					TargetPort: intstr.FromInt(8080),
+				},
+			},
+		},
+	}
+	_, err = client.CoreV1().Services(ns.Name).Create(tCtx, svc, metav1.CreateOptions{})
+	if err != nil {
+		t.Fatalf("Error creating service: %v", err)
+	}
+
+	validateLabelsOnEndpointAndEndpointSlice(t, tCtx, client, ns.Name, svc.Name,
+		[]string{corev1.IsHeadlessService, "app"}, nil)
+
+	svc.Labels["new-label"] = "new-label-value"
+	delete(svc.Labels, "app")
+
+	_, err = client.CoreV1().Services(ns.Name).Update(tCtx, svc, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("Error updating service: %v", err)
+	}
+
+	// Validate that the new labels are propagated.
+	validateLabelsOnEndpointAndEndpointSlice(t, tCtx, client, ns.Name, svc.Name,
+		[]string{corev1.IsHeadlessService, "new-label"}, []string{"app"})
+}
+
+// TestEndpointsHeadlessLabel tests how the service.kubernetes.io/headless
+// label is applied to Endpoints and EndpointSlices based on Service properties.
+func TestEndpointsHeadlessLabel(t *testing.T) {
+	tCtx := ktesting.Init(t)
+
+	// Disable ServiceAccount admission plugin as we don't have serviceaccount controller running.
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, framework.DefaultTestServerFlags(), framework.SharedEtcd())
+	defer server.TearDownFn()
+
+	client, err := clientset.NewForConfig(server.ClientConfig)
+	if err != nil {
+		t.Fatalf("Error creating clientset: %v", err)
+	}
+
+	resyncPeriod := 12 * time.Hour
+	informers := informers.NewSharedInformerFactory(client, resyncPeriod)
+
+	epController := endpoint.NewEndpointController(
+		tCtx,
+		informers.Core().V1().Pods(),
+		informers.Core().V1().Services(),
+		informers.Core().V1().Endpoints(),
+		client,
+		0)
+
+	epsController := endpointslice.NewController(
+		tCtx,
+		informers.Core().V1().Pods(),
+		informers.Core().V1().Services(),
+		informers.Core().V1().Nodes(),
+		informers.Discovery().V1().EndpointSlices(),
+		int32(100),
+		client,
+		1*time.Second)
+
+	// Start informer and controllers
+	informers.Start(tCtx.Done())
+	go epController.Run(tCtx, 1)
+	go epsController.Run(tCtx, 1)
+
+	testCases := []struct {
+		name                string
+		serviceLabels       map[string]string
+		isHeadless          bool
+		expectHeadlessLabel bool
+	}{
+		{
+			name:                "non-headless service with no labels",
+			serviceLabels:       nil,
+			isHeadless:          false,
+			expectHeadlessLabel: false,
```

Comment:
`isHeadless == expectHeadlessLabel` in every case

Reply:
ups